Introduce formal principle documentation to support contrib changes (#3768)
Pyomo/pyomo: doc/OnlineDocs/contribution_guide.rst

closer to 90%. We also require that all tests pass before a PR will be 
merged.

Tests must import the Pyomo test harness from
``pyomo.common.unittest`` instead of using Python's built-in
``unittest`` module directly. This wrapper extends the standard testing
framework with Pyomo-specific capabilities, including automatic plugin
discovery, solver availability checks, and improved test filtering
controls. Using the provided interface ensures that all tests run
consistently across Pyomo's multiple CI environments.
A small example is shown below:

.. code-block:: python

   import pyomo.common.unittest as unittest

   class TestSomething(unittest.TestCase):
       def test_basic(self):
           self.assertEqual(1 + 1, 2)

Developers can also use any of the predefined ``pytest`` markers to categorize
their tests appropriately.
Markers are declared in ``pyproject.toml``. Some commonly used markers are:

- ``expensive``: tests that take a long time to run
- ``mpi``: tests that require MPI
- ``solver(name)``: dynamic marker to label a test for a specific solver,

>   e.g., ``@pytest.mark.solver("gurobi")``

Author: I added a link to the root-dir `conftest.py` which explains the marker application logic we have created in Pyomo.

closer to 90%. We also require that all tests pass before a PR will be 
merged.

Tests must import the Pyomo test harness from
``pyomo.common.unittest`` instead of using Python's built-in
``unittest`` module directly. This wrapper extends the standard testing
framework with Pyomo-specific capabilities, including automatic plugin
discovery, solver availability checks, and improved test filtering

> controls. Using the provided interface ensures that all tests run

Reviewer: This is not technically true (the availability / plugin / filtering from `conftest.py` has nothing to do with Pyomo's custom unittest module.
```suggestion
framework with Pyomo-specific capabilities, including support for test
timeouts and Pyomo-specific assertions for comparing expressions and
nested containers with numerical tolerance.
Using the provided interface ensures that all tests run
```

After a PR is opened it will be reviewed by at least two members of the
core development team. The core development team consists of anyone with
write-access to the Pyomo repository. PRs opened by a core
developer only require one review. The reviewers will decide if they
think a PR should be merged or if more changes are necessary.

Reviewers look for:

* **Core and Addons:** Code rigor, standards compliance, appropriate test
  coverage, and avoidance of unintended side effects
* **Devel:** Basic code correctness and clarity, with an understanding that
  these areas are experimental and evolving

> * **All areas:** Reasonable documentation and tests

Reviewer 2: ```suggestion
* **All areas:** Code formatting (using `black`), documentation, and tests
```

in. Once you have this branch checked out, you can start coding. Bug
fixes and minor enhancements to existing Pyomo functionality should be
made in the appropriate files in the Pyomo code base.

Larger features, new modeling components, or experimental functionality
should be placed in one of Pyomo's extension namespaces, described below.

Namespaces for Contributed and Experimental Code
++++++++++++++++++++++++++++++++++++++++++++++++

Pyomo has a long history of supporting community-developed extensions.
Historically, all such contributions were placed under the
``pyomo.contrib`` namespace. This structure allowed new modeling tools and
algorithms to be shared quickly, but over time it became difficult for users
to distinguish between more stable, supported functionality and
experimental or research-oriented code.

As a result, Pyomo is transitioning to a more structured contribution
model with two clear namespaces:

* ``pyomo.addons`` – For mostly stable, supported extensions that build on
  the Pyomo core. These packages are maintained by dedicated
  contributors, follow Pyomo's coding and testing standards, and adhere

>   to the same deprecation policies as the rest of the codebase.

Reviewer 2: ```suggestion
  to the same backwards compatibility and deprecation policies as the rest of the codebase.
```

* ``pyomo.devel`` – For experimental or rapidly evolving
  contributions. These modules serve as early experimentation for research ideas,
  prototypes, or specialized modeling components. Functionality under
  this namespace may change or be removed between releases without
  deprecation warnings.

This two-tiered structure provides contributors a clear pathway from
**experimentation to supported integration**, while protecting users from
unexpected changes in stable areas of the codebase.

.. list-table::
   :header-rows: 1
   :widths: 20 30 50

   * - Namespace
     - Intended Use
     - Stability
   * - ``pyomo.devel``
     - Active research and experimental code
     - Unstable; APIs may change without warning
   * - ``pyomo.addons``
     - Mostly stable, supported extensions maintained by contributors
     - Mostly stable APIs; follow Pyomo's standards
   * - ``pyomo``
     - Core Pyomo modeling framework
     - Fully supported and versioned

For specific inclusion requirements and maintenance expectations for each
namespace, see:

* :doc:`../../pyomo/addons/README`

> * :doc:`../../pyomo/devel/README`

Author: I am woefully unaware of the right way to get links to files "in the codebase but not in the doc tree" to render. I think the smartest choice is to link to the files on GitHub, but they don't exist yet. So we're in a bit of a catch-22 here. I think I am just going to _not_ link and instead say, "see each directory's accompanying README"

namespace, see each directory's accompanying ``README.md``.

Submitting a Contributed Package
--------------------------------

Including contributed packages in the Pyomo source tree provides a
convenient mechanism for defining new functionality that can be
optionally deployed by users. We expect this mechanism to include
Pyomo extensions and experimental modeling capabilities.  However,
contributed packages are treated as optional packages, which are not
maintained by the Pyomo developer team. Thus, it is the responsibility
of the code contributor to keep these packages up-to-date.

Contributed packages will be considered as pull requests,
which will be reviewed by the Pyomo developer team. Specifically,
this review will consider the suitability of the proposed capability,
whether tests are available to check the execution of the code, and
whether documentation is available to describe the capability.
Contributed packages will be tested along with Pyomo. If test failures
arise, then these packages will be disabled and an issue will be
created to resolve these test failures. The Pyomo team reserves the
right to remove contributed packages that are not maintained.

> When submitting a new contributed package (under either ``addons`` or

Reviewer: ```suggestion
When submitting a new package (under either ``addons`` or
```

``devel``), please ensure that:

* The package has at least one maintainer responsible for its upkeep.
* The code includes tests that can be run through Pyomo's
  continuous integration framework.
* The package includes a README or module-level documentation that

>   clearly describes its purpose and usage.

Reviewer 3: Is this necessary if docs are added to the online docs?

* Optional dependencies are properly declared in ``setup.py``
  under the appropriate ``[optional]`` section.
* The contribution passes all standard style and formatting checks.

> 

Reviewer 4: Oh, just kidding, you address this below. Perhaps it's worth including *where* to go to get the imports right though.

Contributed Packages within Pyomo
+++++++++++++++++++++++++++++++++

> Third-party contributions can be included directly within the

Reviewer: I would like to remove all references / occurrences of "third party"

Namespaces for Contributed and Experimental Code
++++++++++++++++++++++++++++++++++++++++++++++++

Pyomo organizes non-core functionality into a small
number of clearly defined namespaces. Contributors should place new
functionality according to its intended stability and maintenance
expectations:

* ``pyomo.addons`` – For mostly stable, supported extensions that build on
  the Pyomo core. These packages are maintained by dedicated
  contributors, follow Pyomo's coding and testing standards, and adhere
  to the same backwards compatibility and deprecation policies as the
  rest of the codebase.

* ``pyomo.devel`` – For experimental or rapidly evolving
  contributions. These modules serve as early experimentation for research ideas,
  prototypes, or specialized modeling components. Functionality under
  this namespace may change or be removed between releases without
  deprecation warnings.

* ``pyomo.unsupported`` - For contributions that no longer have an active
  maintainer nor any future development plans. Functionality under this namespace
  may not work and is **NOT** tested through the standard test harness.

> This three-tiered structure provides contributors a clear pathway from

Reviewer 2: ```suggestion
This tiered namespace structure provides contributors a clear pathway from
```
(because counting is hard and I couldn't decide if it should be 3 or 4)

Core Functionality
++++++++++++++++++

The main functionality of the contributed package should be implemented
in one or more modules within the package directory (for example,
``core.py``). These modules should follow Pyomo's coding standards,
documentation requirements, and dependency management policies.

Tests
+++++

All contributed packages must include tests. Tests should be placed in a
``tests`` subpackage and use the Pyomo test harness provided by
``pyomo.common.unittest``.

At a minimum, tests should verify that the package can be imported and
that its primary functionality executes as expected. For example::

   import pyomo.common.unittest as unittest

   class TestExamplePackage(unittest.TestCase):
       def test_import(self):
           import pyomo.devel.example_package

Tests for contributed packages are run alongside the rest of the Pyomo
test suite and must not require optional dependencies unless those
dependencies are properly declared and tests are appropriately marked.
For example, tests that require ``numpy`` may be marked using the Pyomo
test harness as follows::

   import pyomo.common.unittest as unittest

>    from pyomo.common.dependencies import numpy_available

Reviewer: ```suggestion
   from pyomo.common.dependencies import numpy as np, numpy_available
```
